File: src/main/java/com/tealcube/minecraft/bukkit/mythicdrops/api/enchantments/MythicEnchantment.java
/*
 * The MIT License
 * Copyright © 2013 Richard Harrah
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.tealcube.minecraft.bukkit.mythicdrops.api.enchantments;

import com.tealcube.minecraft.bukkit.mythicdrops.utils.RandomRangeUtil;
import java.util.Objects;
import org.apache.commons.lang3.math.NumberUtils;
import org.bukkit.enchantments.Enchantment;

/**
 * A class containing an {@link Enchantment} and a minimum and maximum level.
 */
public final class MythicEnchantment {

  private final Enchantment enchantment;
  private final int minimumLevel;
  private final int maximumLevel;

  /**
   * Instantiate a new MythicEnchantment with an {@link Enchantment} and a minimum and maximum level.
   *
   * @param enchantment Enchantment to use
   * @param minimumLevel minimum level of Enchantment
   * @param maximumLevel maximum level of Enchantment
   */
  public MythicEnchantment(Enchantment enchantment, int minimumLevel, int maximumLevel) {
    this.enchantment = enchantment;
    this.minimumLevel = Math.max(1, Math.min(minimumLevel, maximumLevel));
    this.maximumLevel = Math.min(Math.max(minimumLevel, maximumLevel), 127);
  }

  public static MythicEnchantment fromString(String string) {
    Enchantment ench = null;
    int value1 = 0;
    int value2 = 0;
    String[] split = string.split(":");
    switch (split.length) {
      case 0:
        break;
      case 1:
        break;
      case 2:
        ench = Enchantment.getByName(split[0]);
        if (ench == null) {
          break;
        }
        value1 = value2 = NumberUtils.toInt(split[1], 1);
        break;
      default:
        ench = Enchantment.getByName(split[0]);
        if (ench == null) {
          break;
        }
        value1 = NumberUtils.toInt(split[1], 1);
        value2 = NumberUtils.toInt(split[2], 1);
        break;
    }
    if (ench == null) {
      return null;
    }
    return new MythicEnchantment(ench, value1, value2);
  }

  /**
   * Gets the {@link Enchantment}.
   *
   * @return Enchantment
   */
  public Enchantment getEnchantment() {
    return enchantment;
  }

  /**
   * Returns the range between the minimum and maximum levels of the Enchantment. <br> Equivalent of {@code {@link
   * #getMaximumLevel()} - {@link #getMinimumLevel()} }
   *
   * @return range between the minimum and maximum levels
   */
  public double getRange() {
    return getMaximumLevel() - getMinimumLevel();
  }

  /**
   * Returns the maximum level of the Enchantment.
   *
   * @return maximum level
   */
  public int getMaximumLevel() {
    return Math.min(maximumLevel, 127);
  }

  /**
   * Returns the minimum level of the Enchantment.
   *
   * @return minimum level
   */
  public int getMinimumLevel() {
    return Math.max(minimumLevel, 1);
  }

  public int getRandomLevel() {
    return RandomRangeUtil.randomRange(getMinimumLevel(), getMaximumLevel());
  }

  @Override
  public String toString() {
    if (enchantment != null) {
      return String.format("%s:%s:%s", enchantment.getName(), minimumLevel, maximumLevel);
    } else {
      return null;
    }
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    MythicEnchantment that = (MythicEnchantment) o;
    return minimumLevel == that.minimumLevel &&
        maximumLevel == that.maximumLevel &&
        Objects.equals(enchantment, that.enchantment);
  }

  @Override
  public int hashCode() {
    return Objects.hash(enchantment, minimumLevel, maximumLevel);
  }
}

File: src/main/kotlin/com/tealcube/minecraft/bukkit/mythicdrops/api/enchantments/MythicEnchantment.kt
/*
 * This file is part of MythicDrops, licensed under the MIT License.
 *
 * Copyright (C) 2019 Richard Harrah
 *
 * Permission is hereby granted, free of charge,
 * to any person obtaining a copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.tealcube.minecraft.bukkit.mythicdrops.api.enchantments

import com.tealcube.minecraft.bukkit.mythicdrops.utils.EnchantmentUtil
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.enchantments.Enchantment
import kotlin.math.max
import kotlin.math.min

/**
 * Represents an enchantment with a minimum level and a maximum level.
 *
 * @property enchantment Minecraft enchantment to wrap
 * @property minimumLevel Minimum level of enchantment
 * @property maximumLevel Maximum level of enchantment
 */

data class MythicEnchantment(
    val enchantment: Enchantment,
    internal val pMinimumLevel: Int,
    internal val pMaximumLevel: Int = pMinimumLevel,
    val minimumLevel: Int = min(pMinimumLevel, pMaximumLevel).coerceAtLeast(1),
    val maximumLevel: Int = max(pMinimumLevel, pMaximumLevel).coerceAtMost(HIGHEST_ENCHANTMENT_LEVEL)
) {
    companion object {
        const val HIGHEST_ENCHANTMENT_LEVEL = 127
        private const val DEFAULT_ENCHANTMENT_LEVEL = 1
        private val colonRegex = """:""".toRegex()

        /**
         * Constructs a [MythicEnchantment] from a [ConfigurationSection] and its associated [key].
         *
         * @param configurationSection ConfigurationSection of a YAML file
         * @param key Key for the given [configurationSection]
         *
         * @return Constructed MythicEnchantment, or null if unable to find matching [Enchantment]
         */
        @JvmStatic
        fun fromConfigurationSection(
            configurationSection: ConfigurationSection,
            key: String
        ): MythicEnchantment? {
            val enchantment = EnchantmentUtil.getByKeyOrName(key) ?: return null
            val minimumLevel =
                max(
                    configurationSection.getInt("$key.minimumLevel", DEFAULT_ENCHANTMENT_LEVEL),
                    configurationSection.getInt("$key.minimum-level", DEFAULT_ENCHANTMENT_LEVEL)
                )
            val maximumLevel =
                max(
                    configurationSection.getInt("$key.maximumLevel", DEFAULT_ENCHANTMENT_LEVEL),
                    configurationSection.getInt("$key.maximum-level", DEFAULT_ENCHANTMENT_LEVEL)
                )
            return MythicEnchantment(enchantment, minimumLevel, maximumLevel)
        }

        /**
         * Constructs a [MythicEnchantment] from a [String] in the format "enchantment:minimumLevel:maximumLevel".
         *
         * Also accepts "enchantment:level" format.
         *
         * @param string String in "enchantment:minimumLevel:maximumLevel" format.
         * @return Constructed MythicEnchantment, or null if unable to find matching [Enchantment]
         */
        @JvmStatic
        fun fromString(string: String): MythicEnchantment? {
            var enchantment: Enchantment? = null
            var value1 = 0
            var value2 = 0
            val split = string.split(colonRegex).dropLastWhile { it.isEmpty() }.toTypedArray()
            when (split.size) {
                0 -> {
                }

                1 -> {
                }

                2 -> {
                    enchantment = EnchantmentUtil.getByKeyOrName(split[0])
                    if (enchantment != null) {
                        value1 = split[1].toIntOrNull() ?: DEFAULT_ENCHANTMENT_LEVEL
                        value2 = value1
                    }
                }

                else -> {
                    enchantment = EnchantmentUtil.getByKeyOrName(split[0])
                    if (enchantment != null) {
                        value1 = split[1].toIntOrNull() ?: DEFAULT_ENCHANTMENT_LEVEL
                        value2 = split[2].toIntOrNull() ?: DEFAULT_ENCHANTMENT_LEVEL
                    }
                }
            }
            return if (enchantment == null) {
                null
            } else {
                MythicEnchantment(enchantment, value1, value2)
            }
        }
    }

    /**
     * Gets a random level value between [minimumLevel] (inclusive) and [maximumLevel] (inclusive).
     *
     * @return random level
     */
    fun getRandomLevel() =
        if (minimumLevel < maximumLevel) {
            (minimumLevel..maximumLevel).random()
        } else {
            minimumLevel
        }

    override fun toString(): String {
        return "$enchantment:$minimumLevel:$maximumLevel"
    }
}
